## 问题一

In [9]:
#加载头文件
import numpy as np
import pandas as pd
import scipy.stats as stats
import statistics as sta
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns

In [10]:
x=np.array([1200, 400, -200, 800, 600, 300, 1000, -100, 500, 700, 200, 900, -300, 100, 400, 650, 550, -150, 300, 750])
mean=x.mean()
std=x.std(ddof=1)
print("样本均值为：",round(mean,2))
print("样本标准差为：",round(std,2))

样本均值为： 430.0
样本标准差为： 416.25


In [11]:
ci=0.95

def mean_ci_t(data,ci):
    n=len(data)
    std=np.std(data,ddof=1)
    m=np.mean(data)
    sem=std/np.sqrt(n)
    moe=stats.t.isf((1-ci)/2,n-1)*sem
    return m-moe,m+moe

l,h=mean_ci_t(x,ci)
print("均值的95%置信区间是: [{:.2f}, {:.2f}]".format(l,h))


均值的95%置信区间是: [235.19, 624.81]


In [12]:
alpha=0.05
mu=200
def right_mean_t(data,alpha,mu):
    n=len(data)
    std=np.std(data,ddof=1)
    m=np.mean(data)
    sem=std/np.sqrt(n)
    t_obs=(m-mu)/sem
    p=stats.t.sf(t_obs,n-1)
    return p,t_obs

p,t=right_mean_t(x,alpha,mu)
print("t统计量为：",round(t,2))
print("p值为：",round(p,4))

t统计量为： 2.47
p值为： 0.0116


1. 样本均值为： 430.0，样本标准差为： 416.25

2. 均值的95%置信区间是: [235.19, 624.81]。200不在置信区间内，所以不能判断高原集训的平均增益可能只是 200 ml 左右。

3. t统计量为： 2.47，p值为： 0.0116。$P < \alpha$，所以拒绝零假设，故在 $\alpha$=0.05的条件下，高原集训的平均收益大于200ml

4. 可以看到CI和NHST的结论是一致的：95%置信区间 [235.19, 624.81] 不包含200；=0.05的假设检验也拒绝H0，结论完全一致。

## 问题二

In [17]:
a=np.array([12, 15, 8, 10, 18, 9, 14, 11, 7, 13, 16, 10])
b=np.array([14, 17, 9, 12, 20, 11, 13, 15, 10, 16, 18, 12])
m_a=a.mean()
m_b=b.mean()
std_a=a.std(ddof=1)
std_b=b.std(ddof=1)
print("A组样本均值为：",round(m_a,2),"标准差为：",round(std_a,2))
print("B组样本均值为：",round(m_b,2),"标准差为：",round(std_b,2))

A组样本均值为： 11.92 标准差为： 3.37
B组样本均值为： 13.92 标准差为： 3.37


In [14]:
ci=0.95

def double_mean_ci_t(data1,data2,ci):
    n1=len(data1)
    n2=len(data2)
    std1=np.std(data1,ddof=1)
    std2=np.std(data2,ddof=1)
    m1=np.mean(data1)
    m2=np.mean(data2)
    m=m2-m1
    var1=std1**2/n1
    var2=std2**2/n2
    sem=np.sqrt(var1+var2)
    df=(var1+var2)**2/(var1**2/(n1-1)+var2**2/(n2-1))
    moe=stats.t.isf((1-ci)/2,df)*sem
    return m-moe,m+moe

l,h=double_mean_ci_t(a,b,ci)
print("均值差的95%置信区间是: [{:.2f}, {:.2f}]".format(l,h))

均值差的95%置信区间是: [-0.85, 4.85]


In [15]:
def double_mean_t(data1, data2):
    n1=len(data1)
    n2=len(data2)
    std1=np.std(data1,ddof=1)
    std2=np.std(data2,ddof=1)
    m1=np.mean(data1)
    m2=np.mean(data2)
    m=m2-m1
    var1=std1**2/n1
    var2=std2**2/n2
    sem=np.sqrt(var1+var2)
    df=(var1+var2)**2/(var1**2/(n1-1)+var2**2/(n2-1))
    t_obs=m/sem
    p=2 * stats.t.sf(abs(t_obs),df)
    return p,t_obs

p,t=double_mean_t(a,b)
print("t统计量为：",round(t,2))
print("p值为：",round(p,4))

t统计量为： 1.45
p值为： 0.1601


1. A组样本均值为： 11.92 标准差为： 3.37；B组样本均值为： 13.92 标准差为： 3.37。
2. 均值差的95%置信区间是: [-0.85, 4.85]
3. t统计量为： 1.45，p值为： 0.1601。$p > \alpha$，所以不拒绝零假设，故在 $\alpha$=0.05的条件下，可以认为两组的差异是不显著的。
4. 95%置信区间是[-0.85, 4.85]，包含在 ±8 mmHg 的范围内，可以支持“B 与 A 在 ±8 mmHg 范围内等价” 。零假设检验只能证伪，不拒绝零假设只能代表在当前显著性水平下，根据提供的样本不能推翻H0，代表没有足够的证据来证明H0，等价需要有足够的证据，所以不代表完全等价。